File: neet_code_all_cpp/math_and_geometry/count_odd_number_in_interval.cpp
#include <iostream>

class Solution {
    public:
        // AC
        int countOdds(int low, int high) {
            int num = high - low + 1;
            if(num % 2 == 0) {
                return num / 2;
            }

            if(low % 2 == 0) {
                return num / 2;
            } else {
                return num / 2 + 1;
            }
        }

        // AC
        int countOdds2(int low, int high) {
            int num = high - low + 1;
            if(num % 2 == 0 || low % 2 == 0) {
                return num / 2;
            }

            return num / 2 + 1;
        }
};

// Kotlinの模範解答より
class SolutionAnsKotlin {
    public:
        // AC
        int countOdds1(int low, int high) {
            int result = 0;
            result += (high - low) / 2;
            if(low % 2 == 1 || high % 2 == 1) {
                ++result;
            }

            return result;
        }

        // AC
        int countOdds2(int low, int high) {
            return (high + 1) / 2 - low / 2;
        }
};

int main(void) {
    std::pair<int, int> case_1 = {3, 7};
    std::pair<int, int> case_2 = {8, 10};

    Solution s_1;

    std::cout << s_1.countOdds(case_1.first, case_1.second) << std::endl;
    std::cout << s_1.countOdds(case_2.first, case_2.second) << std::endl;

    std::cout << s_1.countOdds2(case_1.first, case_1.second) << std::endl;
    std::cout << s_1.countOdds2(case_2.first, case_2.second) << std::endl;

    SolutionAnsKotlin s_ans_kt;

    std::cout << s_ans_kt.countOdds1(case_1.first, case_1.second) << std::endl;
    std::cout << s_ans_kt.countOdds1(case_2.first, case_2.second) << std::endl;

    std::cout << s_ans_kt.countOdds2(case_1.first, case_1.second) << std::endl;
    std::cout << s_ans_kt.countOdds2(case_2.first, case_2.second) << std::endl;
}

File: neet_code_all/src/math_and_geometry/count_odd_number_in_interval.rs
struct Solution;
impl Solution {
    // TLE
    pub fn count_odds_linear(low: i32, high: i32) -> i32 {
        let mut result = 0;
        for i in low..=high {
            if i % 2 == 1 {
                result += 1;
            }
        }

        result
    }

    // AC
    pub fn count_odds(low: i32, high: i32) -> i32 {
        let num = high - low + 1;
        if num % 2 == 0 {
            return num / 2;
        }

        if low % 2 == 0 {
            num / 2
        } else {
            num / 2 + 1
        }
    }

    // AC
    pub fn count_odds_2(low: i32, high: i32) -> i32 {
        let num = high - low + 1;
        if num % 2 == 0 || low % 2 == 0 {
            return num / 2;
        }

        num / 2 + 1
    }
}

// Kotlinの模範解答より
struct SolutoinAnsKotlin;
impl SolutoinAnsKotlin {
    // AC
    pub fn count_odds_1(low: i32, high: i32) -> i32 {
        let mut result = 0;
        result += (high - low) / 2;
        if low % 2 == 1 || high % 2 == 1 {
            result += 1;
        }

        result
    }

    // AC
    pub fn count_odds_2(low: i32, high: i32) -> i32 {
        (high + 1) / 2 - low / 2
    }
}

fn main() {
    let case_1 = (3, 7);
    let case_2 = (8, 10);

    println!(
        "case_1: {}",
        Solution::count_odds_linear(case_1.0, case_1.1)
    );
    println!(
        "case_2: {}",
        Solution::count_odds_linear(case_2.0, case_2.1)
    );

    println!("case_1: {}", Solution::count_odds(case_1.0, case_1.1));
    println!("case_2: {}", Solution::count_odds(case_2.0, case_2.1));

    println!("case_1: {}", Solution::count_odds_2(case_1.0, case_1.1));
    println!("case_2: {}", Solution::count_odds_2(case_2.0, case_2.1));

    println!(
        "case_1: {}",
        SolutoinAnsKotlin::count_odds_1(case_1.0, case_1.1)
    );
    println!(
        "case_2: {}",
        SolutoinAnsKotlin::count_odds_1(case_2.0, case_2.1)
    );

    println!(
        "case_1: {}",
        SolutoinAnsKotlin::count_odds_2(case_1.0, case_1.1)
    );
    println!(
        "case_2: {}",
        SolutoinAnsKotlin::count_odds_2(case_2.0, case_2.1)
    );
}
